File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/core/action/Action.java
package org.firstinspires.ftc.teamcode.core.action;

import java.util.List;
import java.util.Set;

import org.firstinspires.ftc.teamcode.core.Module;

/**
 * Composable action: a sequence of steps the {@link Actions} scheduler ticks cooperatively
 * on the OpMode thread. Construct via {@link ActionBuilder} or {@link Actions} factories.
 */
public final class Action {

    final List<Step> steps;
    private final Set<Module> targets;
    private final String name;

    private int currentStepIndex = -1;
    private boolean stepStarted = false;
    private boolean done = false;
    private boolean cancelled = false;
    private boolean completedEarly = false;
    private boolean embedded = false;

    Action(List<Step> steps, Set<Module> targets, String name) {
        this.steps = steps;
        this.targets = targets;
        this.name = name;
    }

    /** Schedule; conflicting actions (shared target module) are cancelled. Returns immediately. */
    public void schedule() {
        Actions.schedule(this);
    }

    public void cancel() {
        if (done || cancelled) return;
        cancelled = true;
        Step current = currentStepIfRunning();
        if (current != null) current.cancel(this);
    }

    public boolean isRunning() {
        return !done && !cancelled && currentStepIndex != -1 && Actions.contains(this);
    }

    public boolean isComplete() {
        return done && !cancelled;
    }

    public boolean isCancelled() {
        return cancelled;
    }

    /**
     * Marked once this Action's steps/children have been folded into another action (composed via
     * {@code action()}, {@code sequence}, {@code parallel}, {@code race}, etc.). Such an action shares
     * mutable Step state with its host, so scheduling it independently would corrupt both — the
     * scheduler refuses it. Build a fresh action per use instead.
     */
    void markEmbedded() { embedded = true; }
    boolean isEmbedded() { return embedded; }

    public float getProgress() {
        if (steps.isEmpty() || done) return 1f;
        if (currentStepIndex < 0) return 0f;
        return (float) currentStepIndex / steps.size();
    }

    public Set<Module> getTargets() {
        return targets;
    }

    public String getName() {
        return name;
    }

    /** Stop after the current step finishes without flagging the action as cancelled. */
    public void completeEarly() {
        completedEarly = true;
    }

    void reset() {
        currentStepIndex = -1;
        stepStarted = false;
        done = false;
        cancelled = false;
        completedEarly = false;
    }

    /**
     * Advance one scheduler tick. Instant steps chain within a single tick — only yield when a
     * step returns false from {@link Step#tick}.
     */
    boolean tick() {
        if (done || cancelled) return true;

        while (true) {
            // A step's start/tick may have indirectly cancelled us (e.g. via scheduled conflict).
            if (done || cancelled) return true;
            if (completedEarly) {
                done = true;
                return true;
            }
            if (currentStepIndex == -1) {
                currentStepIndex = 0;
                stepStarted = false;
            }
            if (currentStepIndex >= steps.size()) {
                done = true;
                return true;
            }
            Step step = steps.get(currentStepIndex);
            if (!stepStarted) {
                try {
                    step.start(this);
                } catch (Exception e) {
                    logStepFailure(currentStepIndex, e);
                    try { step.cancel(this); } catch (Exception ignored) {}
                    cancelled = true;
                    return true;
                }
                stepStarted = true;
            }
            boolean stepDone;
            try {
                stepDone = step.tick(this);
            } catch (Exception e) {
                logStepFailure(currentStepIndex, e);
                try { step.cancel(this); } catch (Exception ignored) {}
                cancelled = true;
                return true;
            }
            if (!stepDone) return false;
            currentStepIndex++;
            stepStarted = false;
        }
    }

    private Step currentStepIfRunning() {
        if (currentStepIndex < 0 || currentStepIndex >= steps.size()) return null;
        return steps.get(currentStepIndex);
    }

    private void logStepFailure(int stepIndex, Exception e) {
        StringBuilder info = new StringBuilder();
        if (targets.isEmpty()) {
            info.append("no targets");
        } else {
            info.append("targets: [");
            int i = 0;
            for (Module m : targets) {
                if (i++ > 0) info.append(", ");
                info.append(m.getName());
            }
            info.append(']');
        }
        System.err.println("[Action] '" + name + "' failed at step "
                + (stepIndex + 1) + "/" + steps.size() + " (" + info + "): " + e.getMessage());
        e.printStackTrace();
    }

    public Action then(Action other) { return Actions.sequence(this, other); }
    public Action with(Action other) { return Actions.parallel(this, other); }
    public Action timeout(long ms) { return Actions.timeout(this, ms); }
    public Action withName(String newName) { return new Action(steps, targets, newName); }

    @Override
    public String toString() {
        return name + "(" + steps.size() + " steps)";
    }

    /**
     * One unit of work. {@link #start} runs once on activation, {@link #tick} is polled until
     * true, {@link #cancel} fires if the parent action is cancelled mid-step.
     */
    public abstract static class Step {
        protected void start(Action parent) {}
        protected abstract boolean tick(Action parent);
        protected void cancel(Action parent) {}
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/core/action/Action.kt
package org.firstinspires.ftc.teamcode.core.action

import org.firstinspires.ftc.teamcode.core.Module
import org.firstinspires.ftc.teamcode.core.state.State
import java.lang.Runnable
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancelChildren
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicInteger

class Action internal constructor(
    internal val steps: List<Step>,
    private val targetModules: Set<Module>,
    private val actionName: String,
) {
    private var job: Job? = null
    private val stepIndex = AtomicInteger(0)
    private val cancelled = AtomicBoolean(false)
    private val completedEarly = AtomicBoolean(false)

    /** Schedule async; actions targeting the same modules auto-cancel each other. */
    fun run() {
        Actions.run(this)
    }

    /** Block the caller until this action completes. Don't call from the OpMode loop thread. */
    fun runBlocking() {
        runBlocking { runSuspending() }
    }

    suspend fun runSuspending() {
        reset()
        runSteps()
    }

    fun cancel() {
        cancelled.set(true)
        job?.cancel()
    }

    val isRunning: Boolean
        get() = job?.isActive == true

    val isComplete: Boolean
        get() = !isRunning && !cancelled.get() && (completedEarly.get() || stepIndex.get() >= steps.size)

    val isCancelled: Boolean
        get() = cancelled.get()

    val progress: Float
        get() = if (steps.isEmpty()) 1f else stepIndex.get().toFloat() / steps.size

    val targets: Set<Module>
        get() = targetModules

    val name: String
        get() = actionName

    /** Stop after the current step finishes, without flagging the action as cancelled. */
    fun completeEarly() {
        completedEarly.set(true)
    }

    internal fun reset() {
        stepIndex.set(0)
        cancelled.set(false)
        completedEarly.set(false)
        job = null
    }

    internal fun setJob(job: Job) {
        this.job = job
    }

    internal suspend fun runSteps() {
        stepIndex.set(0)

        for ((index, step) in steps.withIndex()) {
            if (cancelled.get() || completedEarly.get()) break
            stepIndex.set(index)
            try {
                step.run(this)
            } catch (e: CancellationException) {
                cancelled.set(true)
                throw e
            } catch (e: Exception) {
                val targetInfo = if (targetModules.isNotEmpty()) {
                    "targets: [" + targetModules.joinToString(", ") { it.name } + "]"
                } else {
                    "no targets"
                }
                System.err.println("[ERROR] Action '$actionName' failed at step ${index + 1}/${steps.size} ($targetInfo): ${e.message}")
                e.printStackTrace()
                break
            }
        }

        if (!cancelled.get()) {
            stepIndex.set(steps.size)
        }
    }

    infix fun then(other: Action): Action = Actions.sequence(this, other)

    infix fun with(other: Action): Action = Actions.parallel(this, other)

    fun timeout(ms: Long): Action = Actions.timeout(this, ms)

    fun named(newName: String): Action = Action(steps, targetModules, newName)

    override fun toString(): String = "$actionName(${steps.size} steps)"

    fun interface Step {
        suspend fun run(parent: Action)
    }
}

object Actions {
    private val active = ConcurrentHashMap<Action, Job>()
    private val scope = CoroutineScope(Dispatchers.Default + SupervisorJob())

    @JvmStatic
    fun set(vararg states: State): Action {
        val b = ActionBuilder()
        states.forEach { b.set(it) }
        return b.build()
    }

    /** Resolve states at execution time rather than build time. */
    @SafeVarargs
    @JvmStatic
    fun setLazy(vararg suppliers: java.util.function.Supplier<out State>): Action =
        ActionBuilder().setLazy(*suppliers).build()

    @JvmStatic
    fun run(code: Runnable): Action = ActionBuilder().run(code).build()

    @JvmStatic
    fun run(code: () -> Unit): Action = ActionBuilder().run(code).build()

    @JvmStatic
    fun runSuspending(code: suspend () -> Unit): Action = ActionBuilder().runSuspending(code).build()

    @JvmStatic
    fun delay(ms: Long): Action = ActionBuilder().delay(ms).build()

    @JvmStatic
    fun noop(): Action = ActionBuilder().build()

    @JvmStatic
    fun waitUntil(condition: () -> Boolean): Action = ActionBuilder().waitUntil(condition).build()

    @JvmStatic
    fun waitUntil(condition: () -> Boolean, timeoutMs: Long): Action =
        ActionBuilder().waitUntil(condition, timeoutMs).build()

    @JvmStatic
    fun sequence(vararg actions: Action): Action {
        val b = ActionBuilder()
        actions.forEach { b.action(it) }
        return b.named("Sequence").build()
    }

    @JvmStatic
    fun parallel(vararg actions: Action): Action =
        ActionBuilder().parallel(*actions).named("Parallel").build()

    @JvmStatic
    fun race(vararg actions: Action): Action =
        ActionBuilder().race(*actions).named("Race").build()

    @JvmStatic
    fun timeout(action: Action, ms: Long): Action =
        ActionBuilder().timeout(action, ms).named("Timeout").build()

    @JvmStatic
    fun repeat(action: Action, times: Int): Action =
        ActionBuilder().repeat(action, times).named("Repeat").build()

    @JvmStatic
    fun loop(action: Action, condition: () -> Boolean): Action =
        ActionBuilder().loop(action, condition).named("Loop").build()

    @JvmStatic
    fun ifThen(condition: () -> Boolean, ifTrue: Action): Action =
        ActionBuilder().ifThen(condition, ifTrue).named("IfThen").build()

    @JvmStatic
    fun ifElse(condition: () -> Boolean, ifTrue: Action, ifFalse: Action): Action =
        ActionBuilder().ifElse(condition, ifTrue, ifFalse).named("IfElse").build()

    @JvmStatic
    fun retry(action: Action, maxAttempts: Int, success: () -> Boolean): Action =
        ActionBuilder().retry(action, maxAttempts, success).named("Retry").build()

    @JvmStatic
    fun retry(action: Action, maxAttempts: Int, delayMs: Long, success: () -> Boolean): Action =
        ActionBuilder().retry(action, maxAttempts, delayMs, success).named("Retry").build()

    @JvmStatic
    fun retry(code: Runnable, maxAttempts: Int, success: () -> Boolean): Action =
        ActionBuilder().retry(code, maxAttempts, success).named("Retry").build()

    @JvmStatic
    fun retry(code: Runnable, maxAttempts: Int, delayMs: Long, success: () -> Boolean): Action =
        ActionBuilder().retry(code, maxAttempts, delayMs, success).named("Retry").build()

    @JvmStatic
    fun builder(): ActionBuilder = ActionBuilder()

    internal fun run(action: Action) {
        active.keys.filter { conflicts(it, action) }.forEach {
            it.cancel()
            active.remove(it)
        }

        active[action]?.let {
            action.cancel()
            active.remove(action)
        }

        action.reset()
        val job = scope.launch {
            val myJob = coroutineContext[Job]!!
            try {
                action.runSteps()
            } finally {
                // Only remove if we're still the active job — guard against a re-launch that
                // replaced us between launch and finally.
                active.remove(action, myJob)
            }
        }

        action.setJob(job)
        active[action] = job
    }

    private fun conflicts(a: Action, b: Action): Boolean =
        a.targets.any { it in b.targets }

    @JvmStatic
    fun cancelAll() {
        active.keys.forEach { it.cancel() }
        active.clear()
    }

    @JvmStatic
    fun cancelFor(vararg modules: Module) {
        active.keys.filter { action -> modules.any { it in action.targets } }.forEach {
            it.cancel()
            active.remove(it)
        }
    }

    @JvmStatic
    fun hasActive(): Boolean = active.isNotEmpty()

    @JvmStatic
    fun activeCount(): Int = active.size

    @JvmStatic
    fun isModuleActive(module: Module): Boolean =
        active.keys.any { module in it.targets }

    @JvmStatic
    fun getActive(): List<Action> = active.keys.toList()

    @JvmStatic
    fun shutdown() {
        cancelAll()
        scope.coroutineContext.cancelChildren()
    }

    @JvmStatic
    fun reset() {
        cancelAll()
    }
}
